### ISSUE TITLE
[keyboard navigation] have to start over with keyboard nav if you open new tab

### ISSUE THREAD

**Issue by lmulhall-phet** [phetsims/friction#50]
**Test device:**

MacBook Air (13-inch, Early 2015)
Hanson

**Operating System:**

macOS High Sierra 10.13.1 (17B1003)
Windows 10

**Browser:**

Chrome Version 62.0.3202.94 (Official Build) (64-bit)
Firefox 57.0.1 (64-bit)
Safari Version 11.0.1 (13604.3.5)

**Problem description:**

Not entirely sure this is an issue, I just wanted to check to see if anyone wanted to make changes...

I noticed that if you use keyboard nav, click on another tab, and then click back into the sim tab, you have to start the keyboard nav process over again, i.e., the sim has no memory of what you were doing with keyboard nav before you clicked into the other tab. I thought this might be slightly irritating for someone who would rather use the keyboard nav. This is for phetsims/QA/issues/75.

Thoughts?

**Steps to reproduce:**

1. Open the sim and mess around with keyboard nav.
2. Open up a new tab or maybe click on one of the links provided by the PhET menu.
3. Click back to the sim tab.

**Screenshots:**

Here's an example where I use the link to the website from the PhET menu:

![squash](https://user-images.githubusercontent.com/31285590/33635781-9fe3f8f8-d9d6-11e7-86b0-af3d3ed77e09.gif)

**Troubleshooting information (do not edit):**

Name: ‪Friction‬
URL: https://www.colorado.edu/physics/phet/dev/html/friction/1.4.0-rc.2/friction_en.html
Version: 1.4.0-rc.2 2017-12-05 18:19:27 UTC
Features missing: touch
User Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.94 Safari/537.36
Language: en-US
Window: 1440x700
Pixel Ratio: 1/1
WebGL: WebGL 1.0 (OpenGL ES 2.0 Chromium)
GLSL: WebGL GLSL ES 1.0 (OpenGL ES GLSL ES 1.0 Chromium)
Vendor: WebKit (WebKit WebGL)
Vertex: attribs: 16 varying: 15 uniform: 1024
Texture: size: 16384 imageUnits: 16 (vertex: 16, combined: 80)
Max viewport: 16384x16384
OES_texture_float: true
Dependencies JSON: {"assert":{"sha":"928741cf","branch":"HEAD"},"axon":{"sha":"d70f5056","branch":"HEAD"},"babel":{"sha":"26366701","branch":"master"},"brand":{"sha":"53cf227b","branch":"HEAD"},"chipper":{"sha":"6f8b4c13","branch":"HEAD"},"dot":{"sha":"f09e44aa","branch":"HEAD"},"friction":{"sha":"d2fa6e36","branch":"HEAD"},"joist":{"sha":"bc8f433b","branch":"HEAD"},"kite":{"sha":"72dcc8ee","branch":"HEAD"},"phet-core":{"sha":"88a51822","branch":"HEAD"},"phetcommon":{"sha":"f0288123","branch":"HEAD"},"query-string-machine":{"sha":"b8bdcc28","branch":"HEAD"},"scenery":{"sha":"31c7d26f","branch":"HEAD"},"scenery-phet":{"sha":"594695db","branch":"HEAD"},"sherpa":{"sha":"0f5f30f7","branch":"HEAD"},"sun":{"sha":"2d4b1ff7","branch":"HEAD"},"tandem":{"sha":"56c77424","branch":"HEAD"}}

**Comment by zepumph** [comment]
@jessegreenberg this sorta seems like an extension of https://github.com/phetsims/scenery/issues/713. We haven't spent any energy on trying to handle these browser/system edge cases. @jessegreenberg should this be dealt with before Friction goes out ya think?

**Comment by jessegreenberg** [comment]
Great question @lmulhall-phet thanks for bringing this up.

@zepumph I agree that this could be an extension of phetsims/scenery#713. I don't believe this needs to block publication of Friction, but it would be good to hear from @terracoda and @emily-phet.

I am also not  sure if this is something that needs to be changed. For me, this only happens when I open a new tab by clicking on a link. If I "ctrl + t" or click on another tab in the browser, Chrome remembers what was previously focused and tab navigation continues there when I go back to the sim. So is this how it is supposed to behave?

**Comment by jessegreenberg** [comment]
Like #52 I think this can be deferred until we discuss throughly in https://github.com/phetsims/a11y-research/issues/69, and it would be OK to deploy this sim. We decide in that issue that we prefer a different behavior but if user navigates to a new tab it seems acceptable that  keyboard navigation would start from the beginning of the sim once user returns to it. @emily-phet @terracoda @zepumph do you agree?

**Comment by emily-phet** [comment]
@jessegreenberg @terracoda @zepumph 
Just for clarification: the issue is that in some browsers, any navigation (keyboard, mouse click, etc.) to a different browser tab results in having to start the keyboard nav sequence over when returning back to the PhET sim tab. Or is the issue more about switching between keyboard and mouse (e.g., mouse click to a new tab and then return to PhET sim results in the restart of nav sequence in the sim, but if you keyboard nav away returning keeps your place in the sim)? 

My understanding of keyboard navigation in a web browser is that we control the sequence of focusable elements within the sim - which ultimately becomes a part of the keyboard sequence within the browser window...so tabbing through the sim and continuing to tab cycles you out and into the browser tabs etc. So navigating to a new tab, and returning to the PhET tab would put the focus on the tab button (or whatever it is) and then into the sim. It makes sense to me that each time you did this, you would "drop into" the PhET sim keyboard focus at the same point - always starting with the first focusable element in the sim's tab order. A benefit of this approach is that once you've done this once you always know where the focus will start in the sim. If you leave the sim and return to it, and the focus is where you left it, you'll have to find it again because it could be anywhere. 

Happy to discuss this further, but for now it seems reasonable to me that each return to the PhET sim starts up the nav sequence at the sim's first focusable element.

**Comment by jessegreenberg** [comment]
> My understanding of keyboard navigation in a web browser is that we control the sequence of focusable elements within the sim - which ultimately becomes a part of the keyboard sequence within the browser window

This is correct. But the browser seems to return focus in different ways depending on how the user left the sim tab. And different browsers do different things. Here is a list of behavior on different browsers:

- Edge/Safari:
  - However you leave the sim tab and then return: focus starts over at top of page.
- IE11:
  - Leaving tab with mouse/keyboard/linkand then returning: Focus *stays* on last active element, you can resume previous keyboard interaction.
- Chrome/Firefox:
  - Leaving tab with mouse/keyboard/link then returning: focus goes to the next focusable item when keyboard nav resumes.

The only exception is links in the PhET Menu. Since the menu disappears after a link is clicked, focus starts over at the top of the page.

So the browser is deciding how it wants to handle focus after user changes tabs. In my opinion, we should let the browser do its thing and not try to force focus to go somewhere atypical for the platform. Eit@emily-phet after reviewing this list, do you have any concerns?

**Comment by emily-phet** [comment]
@jessegreenberg - letting the browser do its thing makes sense to me. We would like the user experience to be consistent with their expectations with a particular browser when it doesn't cause us tremendous work/overhead to allow that.

Do you think this behavior across different browsers is intentional, or just a quirk of how they're each handling focus?

**Comment by jessegreenberg** [comment]
Hmm, I sit corrected. I just tested out  behavior of https://mdn.github.io/learning-area/html/forms/your-first-HTML-form/first-form-styled.html which is a typical HTML form. In all browsers I have tested so far (Chrome, FF, Safari, Edge, IE11), focus goes back to the element that had focus when leaving the tab and interaction can continue. Now I think we should strive for that and figure out why PhET sims behave differently.

I estimate it will take about 1 hour to do this so I think we should include it in this deployment. @emily-phet does that sounds good?

**Comment by zepumph** [comment]
After talking with @jessegreenberg about this problem. We don't think this should block Friction anymore, but instead we should continue on with deployment, and work on issues like this in phetsims/scenery#722. This will hopefully be a long term and robust solution to this problem.

**Comment by zepumph** [comment]
Assigning to @emily-phet to veto or confirm the above comment.

**Comment by jessegreenberg** [comment]
This issue can be closed, we are going to discuss and handle this generally in https://github.com/phetsims/scenery/issues/713 and https://github.com/phetsims/a11y-research/issues/69
